wagtail/admin/auth.py

```diff
@@ -154,13 +155,19 @@ def decorated_view(request, *args, **kwargs):
             return reject_request(request)
 
         if user.has_perms(['wagtailadmin.access_admin']):
+            original_language = None
             if hasattr(user, 'wagtail_userprofile'):
                 language = user.wagtail_userprofile.get_preferred_language()
                 l18n.set_language(language)
-                activate_lang(language)
+                if language != get_language():
+                    original_language = get_language()
+                    activate_lang(language)
                 time_zone = user.wagtail_userprofile.get_current_time_zone()
                 activate_tz(time_zone)
-            return view_func(request, *args, **kwargs)
+            response = view_func(request, *args, **kwargs)
+            if original_language:
+                activate_lang(original_language)
+            return response
```

Comment:
Could we use `django.utils.translation.override` here, as you've done for the `language_preferences` view? It seems a lot neater, and I have more confidence that Django's implementation will handle the edge cases correctly (e.g. what happens if the view throws an exception) than our own code.

Reply:
good point, I'll update ASAP

Reply:
updated